### PR TITLE
Add SemaphoreCI

### PR DESCRIPTION
**Pull request by luismiramirez**
# Add SemaphoreCI

SemaphoreCI is our standard CI engine in all of our projects. We move
this project there to fit with our standards.

# Apply RustFmt and Clippy recommended changes

In order to make the CI work, we need to comply with RustFmt and Clippy
rules.

### CODE REVIEW

**Review by thijsc — APPROVED** [review]

**Review by jeffkreeftmeijer — CHANGES_REQUESTED** [review]
Ah, yes. Good call to run this on CI, indeed!

Could you check the CI failure, so we can merge this while green? :)

**Comment by luismiramirez** [comment]
@jeffkreeftmeijer green achieved!

File: .semaphore/semaphore.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    task:
+      jobs:
+        - name: Formatter and linter
+          commands:
+            - checkout
+            - rustup component add rustfmt clippy
+            - cache restore clippy-${SEMAPHORE_GIT_BRANCH}-${LOCK_CHECKSUM}
+            - cargo fmt --verbose --all -- --check
+            - cargo clippy
+            - cache store clippy-${SEMAPHORE_GIT_BRANCH}-${LOCK_CHECKSUM}
```
Comment (tombruijn) [inline .semaphore/semaphore.yml:20]: This prompts an error. I don't know why it's not failing the build. You need to  specify the path you want to cache.

```
Usage:
      cache store (lookup cachable elements in the project and cache them e.g vendor/bundle)
      cache store [key] [local-path]
Example:
      cache store bundle-$SEMAPHORE_GIT_BRANCH-$(checksum Gemfile.lock) vendor/bundle
! [13:08 16/09/2021]: Incorrect number of arguments!
```

File: .semaphore/semaphore.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      image: registry.semaphoreci.com/rust:1.51
+blocks:
+  - name: 'Default'
+    task:
+      jobs:
+        - name: Formatter and linter
+          commands:
+            - checkout
+            - rustup component add rustfmt clippy
+            - cache restore clippy-${SEMAPHORE_GIT_BRANCH}-${LOCK_CHECKSUM}
```
Comment (tombruijn) [inline .semaphore/semaphore.yml:17]: The `LOCK_CHECKSUM` env var seems to be unset. From the build output:

```
cache restore clippy-${SEMAPHORE_GIT_BRANCH}-${LOCK_CHECKSUM}
MISS: clippy-add-semaphore-ci-
```

File: .semaphore/semaphore.yml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            - cargo fmt --verbose --all -- --check
+            - cargo clippy
+            - cache store clippy-${SEMAPHORE_GIT_BRANCH}-${LOCK_CHECKSUM}
+        - name: Build and tests
+          commands:
+            - checkout
+            - rustup toolchain install nightly
+            - cargo build --verbose
```
Comment (tombruijn) [inline .semaphore/semaphore.yml:25]: I don't think the build step is needed beforehand? `cargo test` will also build it.

**Review by tombruijn — CHANGES_REQUESTED** [review]

File: .semaphore/semaphore.yml (hunk 1 above)
Comment (luismiramirez) [reply .semaphore/semaphore.yml]: I just learned that Rust libraries, as this one, should gitignore the `Cargo.lock` file, so we don't have anything to cache here.

File: .semaphore/semaphore.yml (hunk 2 above)
Comment (luismiramirez) [reply .semaphore/semaphore.yml]: Yes, I forgot to export it with `Cargo.lock` checksum, but it's ignored here. We don't need cache for this build.

**Review by tombruijn — APPROVED** [review]
Just reviewed the Semaphore setup because the rest of the changes are automated.

**Pull request merged**